### PR TITLE
clone.d, remove gotos, 4 useless BlockStatement and 1 switch with 2 cases

### PR DESCRIPTION
**Pull request by ghost**

### CODE REVIEW

File: src/dmd/clone.d [hunk 1]
```diff
@@ -147,19 +147,13 @@ private bool needOpAssign(StructDeclaration sd)
 {
     //printf("StructDeclaration::needOpAssign() %s\n", sd.toChars());
 
-    static bool isNeeded()
```
Comment (WalterBright) [inline src/dmd/clone.d]: This code is here for a reason - so I can hook the function exit.

File: src/dmd/clone.d [hunk 2]
```diff
@@ -419,30 +413,25 @@ bool needOpEquals(StructDeclaration sd)
...
+            return true;
         if (tv.ty == Taarray)
-            goto Lneed;
+            return true;
         if (tv.ty == Tclass)
-            goto Lneed;
+            return true;
     }
-Ldontneed:
-    //printf("\tdontneed\n");
```
Comment (WalterBright) [inline src/dmd/clone.d]: You see, this is why there's a common return.

**Review by thewilsonator — APPROVED** [review]

File: src/dmd/clone.d [hunk 3]
```diff
@@ -739,30 +724,25 @@ private bool needToHash(StructDeclaration sd)
...
```
Comment (WalterBright) [inline src/dmd/clone.d]: Again, this is here for a good reason.

File: src/dmd/clone.d [hunk 4]
```diff
@@ -1137,46 +1117,40 @@ DtorDeclaration buildExternDDtor(AggregateDeclaration ad, Scope* sc)
  */
 FuncDeclaration buildInv(AggregateDeclaration ad, Scope* sc)
 {
-    switch (ad.invs.dim)
-    {
-    case 0:
+    if (ad.invs.dim == 0)
```
Comment (WalterBright) [inline src/dmd/clone.d]: I don't see that anything is really improved by this. It's just a lot of diffs.

File: src/dmd/clone.d [hunk 5]
```diff
@@ -536,16 +525,14 @@ FuncDeclaration buildXopEquals(StructDeclaration sd, Scope* sc)
...
-                Scope scx;
-                /* const bool opEquals(ref const S s);
-                 */
-                auto parameters = new Parameters();
-                parameters.push(new Parameter(STC.ref_ | STC.const_, sd.type, null, null, null));
-                tfeqptr = new TypeFunction(ParameterList(parameters), Type.tbool, LINK.d);
-                tfeqptr.mod = MODFlags.const_;
-                tfeqptr = cast(TypeFunction)tfeqptr.typeSemantic(Loc.initial, &scx);
-            }
+            Scope scx;
```
Comment (WalterBright) [inline src/dmd/clone.d]: this change has some merit, but line 528 should be moved down to just before 535.

File: src/dmd/clone.d [hunk 6]
```diff
@@ -606,16 +593,14 @@ FuncDeclaration buildXopCmp(StructDeclaration sd, Scope* sc)
...
-                Scope scx;
-                /* const int opCmp(ref const S s);
-                 */
-                auto parameters = new Parameters();
-                parameters.push(new Parameter(STC.ref_ | STC.const_, sd.type, null, null, null));
-                tfcmpptr = new TypeFunction(ParameterList(parameters), Type.tint32, LINK.d);
-                tfcmpptr.mod = MODFlags.const_;
-                tfcmpptr = cast(TypeFunction)tfcmpptr.typeSemantic(Loc.initial, &scx);
-            }
+            Scope scx;
```
Comment (WalterBright) [inline src/dmd/clone.d]: This has merit, too.

**Review by WalterBright — CHANGES_REQUESTED** [review]
Keep only the changes I noted as having merit. Thanks.

**Comment by thewilsonator** [comment]
Hmm, looks like it went in anyway. Wonder what happened there?

**Comment by Basile-z** [comment]
GH bug.

File: src/dmd/clone.d (hunk 2 above)
Comment (jacob-carlborg) [reply src/dmd/clone.d]: Wouldn't be be better to use an out contract if you want to examine the function exit?

**Pull request merged**